server/src/main/java/org/elasticsearch/transport/InboundHandler.java

```diff
@@ -335,10 +340,50 @@ private static void sendErrorResponse(String actionName, TransportChannel transp
     private <T extends TransportResponse> void handleResponse(
         InetSocketAddress remoteAddress,
         final StreamInput stream,
-        final TransportResponseHandler<T> handler
+        final TransportResponseHandler<T> handler,
+        final InboundMessage inboundMessage
+    ) {
+        final String executor = handler.executor();
+        if (ThreadPool.Names.SAME.equals(executor)) {
+            // no need to provide a buffer release here, we never escape the buffer when handling directly
+            doHandleResponse(handler, remoteAddress, stream, inboundMessage.getHeader(), () -> {});
+        } else {
+            inboundMessage.incRef();
+            // release buffer once we deserialize the message, but have a fail-safe in #onAfter below in case that didn't work out
+            final Releasable releaseBuffer = Releasables.releaseOnce(inboundMessage::decRef);
```

Comment:
Would be good to have some unit test(s) to demonstrate that we do actually release this. If I make this a no-op then `./gradlew :server:test` still apparently succeeds.

Reply:
It does fail within seconds for `:server:internalClusterTests` though, idk, I never really find much value in these unit tests and they make refactoring this code super hard the more you add of them.
I'd rather not add one here since this is so massively covered by essentially every internal cluster test.

Reply:
Ok nvm should have tried the integrity tests too :)